app/hooks/useNetworkState.tsx

```diff
@@ -0,0 +1,28 @@
+import { useEffect, useRef, useState } from "react";
+
+export function useNetworkState() {
+    const [isOnline, setIsOnline] = useState(window.navigator.onLine);
+    const callbackRef = useRef < () => void> ();
+
+    const on = () => {
+        setIsOnline(true);
+        callbackRef.current && callbackRef.current();
+    };
+    const off = () => setIsOnline(false);
+
+    useEffect(() => {
+        window.addEventListener("online", on);
```

Comment:
What's the browser support like for this feature?

Reply:
From https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine#browser_compatibility, we can see it's well supported.